src/wh_server_she.c

```diff
@@ -119,173 +127,90 @@ static int wh_AesMp16(WOLFHSM_CTX* ctx, byte* in, word32 inSz,
-        /* if erased set to 1 */
-        if (counter == WOLFHSM_ID_ERASED)
-            counter = 1;
-        else
-            counter++;
-        ret = hal_flash_write(writeAddr, (uint8_t*)&counter, sizeof(counter));
-    }
-    /* erase old partition */
-    if (ret == 0)
-        ret = hal_flash_erase(readAddr, WOLFHSM_PARTITION_SIZE);
+    wc_AesFree(sheAes);
     return ret;
 }
 
 /* AuthID is the 4 rightmost bits of messageOne */
-static inline uint16_t hsmShePopAuthId(uint8_t* messageOne)
+static uint16_t hsmShePopAuthId(uint8_t* messageOne)
 {
-    return WOLFHSM_SHE_TRANSLATE_KEY_ID(
-        (*(messageOne + WOLFHSM_SHE_M1_SZ - 1) & 0x0f));
+    return (*(messageOne + WOLFHSM_SHE_M1_SZ - 1) & 0x0f);
 }
 
 /* ID is the second to last 4 bits of messageOne */
-static inline uint16_t hsmShePopId(uint8_t* messageOne)
+static uint16_t hsmShePopId(uint8_t* messageOne)
 {
-    return WOLFHSM_SHE_TRANSLATE_KEY_ID(
-        ((*(messageOne + WOLFHSM_SHE_M1_SZ - 1) & 0xf0) >> 4));
+    return ((*(messageOne + WOLFHSM_SHE_M1_SZ - 1) & 0xf0) >> 4);
 }
 
 /* flags are the rightmost 4 bits of byte 3 as it's leftmost bits
  * and leftmost bit of byte 4 as it's rightmost bit */
-static inline uint32_t hsmShePopFlags(uint8_t* messageTwo)
+static uint32_t hsmShePopFlags(uint8_t* messageTwo)
 {
     return (((messageTwo[3] & 0x0f) << 4) | ((messageTwo[4] & 0x80) >> 7));
 }
 
-int hsmHandleSHE(WOLFHSM_CTX* ctx, wh_Packet* packet,
-    NvmMetaData* meta)
+int wh_Server_HandleSheRequest(whServerContext* server,
+    uint16_t action, uint8_t* data, uint16_t* size)
 {
     int ret = 0;
+    uint32_t i = 0;
     uint32_t field;
+    uint32_t keySz;
     uint8_t* in;
     uint8_t* out;
-    uint8_t* keyOne;
-    uint8_t* keyTwo;
-    uint8_t* keyThree;
-    uint8_t* messageThreeDigest;
+    whPacket* packet = (whPacket*)data;
+    whNvmMetadata meta[1];
     /* TODO we might be able to use the unused part of the packet here to save space since SHE keys are always only 16 bytes */
     uint8_t kdfInput[WOLFHSM_SHE_KEY_SZ * 2];
     uint8_t messageZero[WOLFHSM_SHE_KEY_SZ];
     uint8_t tmpKey[WOLFHSM_SHE_KEY_SZ];
     uint8_t cmacOutput[AES_BLOCK_SIZE];
-    union {
-        Aes aes[1];
-    } crypto;
 
-    switch (packet->subType)
+    if (server == NULL || data == NULL || size == NULL)
+        return WH_ERROR_BADARGS;
+
+    /* TODO does SHE specify what this error should be? */
+    /* if we haven't secure booted, only allow secure boot requests */
+    if ((hsmSheSbState != WOLFHSM_SHE_SB_SUCCESS &&
+        (action != WOLFHSM_SHE_SECURE_BOOT_INIT &&
+        action != WOLFHSM_SHE_SECURE_BOOT_UPDATE &&
+        action != WOLFHSM_SHE_SECURE_BOOT_FINISH &&
+        action != WOLFHSM_SHE_GET_STATUS &&
+        action != WH_SHE_SET_UID)) ||
+        (action != WH_SHE_SET_UID && hsmSheUidSet == 0)) {
+        packet->rc = WH_SHE_ERC_SEQUENCE_ERROR;
+        *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->rc);
+        return 0;
+    }
+
+    switch (action)
```

Comment:
I think we should have handler functions for each individual case - this function is getting pretty gnarly